shared_model/builders/protobuf/block.hpp

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_PROTO_BLOCK_BUILDER_HPP
+#define IROHA_PROTO_BLOCK_BUILDER_HPP
+
+#include "backend/protobuf/block.hpp"
+#include "backend/protobuf/queries/proto_query.hpp"
+#include "backend/protobuf/transaction.hpp"
+#include "block.pb.h"
+
+#include "builders/protobuf/unsigned_proto.hpp"
+#include "interfaces/base/hashable.hpp"
+#include "interfaces/base/signable.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "interfaces/iroha_internal/block.hpp"
+#include "interfaces/transaction.hpp"
+
+namespace shared_model {
+  namespace proto {
+    template <int S = 0>
+    class TemplateBlockBuilder {
+     private:
+      template <class T>
+      using w = detail::PolymorphicWrapper<T>;
+
+      template <int>
+      friend class TemplateBlockBuilder;
+
+      enum RequiredFields {
+        Transactions,
+        TxNumber,
+        Height,
+        PrevHash,
+        CreatedTime,
+        Signatures,
+        TOTAL
+      };
+
+      template <int s>
+      using NextBuilder = TemplateBlockBuilder<S | (1 << s)>;
+
+      iroha::protocol::Block block_;
+
+      template <int Sp>
+      TemplateBlockBuilder(const TemplateBlockBuilder<Sp> &o)
+          : block_(o.block_) {}
+
+     public:
+      TemplateBlockBuilder() = default;
+
+      NextBuilder<Transactions> transactions(
+          const std::vector<w<Transaction>> &transactions) {
+        for (auto tx : transactions) {
+          iroha::protocol::Transaction proto_tx(tx->getTransport());
+          block_.mutable_payload()->mutable_transactions()->Add(std::move(proto_tx));
+        }
+        return *this;
+      }
+
+      NextBuilder<TxNumber> tx_number(
```

Comment:
I think it would be better to use explicit conversion like for transaction/query (and in proposal too).